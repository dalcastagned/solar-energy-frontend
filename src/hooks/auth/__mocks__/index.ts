export const userMock = {
  user: 'teste@teste.com',
  roles: [],
  token:
    'eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJBVVJPUkEiLCJhdWQiOiJBVVJPUkEiLCJleHAiOjMwMjc2NTMzNTAsInJvbGVzIjpbIlBBQ1NOR19ESUNPTV9CVUxLREFUQSIsIlBBQ1NOR19ESUNPTV9XQURPIiwiUEFDU05HX0RJQ09NX1FJRE8iLCJQQUNTTkdfQ09ORklHVVJBVElPTiIsIlBBQ1NOR19OT0RFX0NPTlRST0wiLCJQQUNTTkdfU0VSSUVTX01JR1JBVElPTiIsIlBBQ1NOR19TRVJJRVNfUkVNT1ZBTCJdLCJwcmVmZXRjaHJvbGVzIjpbXSwidXNlciI6InVzZXJ0ZXN0ZSJ9.8a_ZdpxfM1_rU-ekCaZnefGFsJVq2_LetC5yBD_80SE',
};

export const loginRequestMock = {
  email: 'usuario@usuario.com',
  password: 'password',
};

