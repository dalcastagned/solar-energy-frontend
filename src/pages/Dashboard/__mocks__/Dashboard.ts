export const counts = {
  activePlants: 0,
  inactivePlants: 0,
};

export const monthsGenerations = [
  {
    month: 'janeiro',
    generation: 150,
  },
];
