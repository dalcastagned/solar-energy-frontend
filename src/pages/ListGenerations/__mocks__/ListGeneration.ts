export const generationsMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 1,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  generations: [
    {
      id: 1,
      date: '2022-06-05T17:53:08.189Z',
      generatePower: 150,
    },
  ],
};

export const generationMock = {
  id: 1,
  date: '2022-06-05T17:53:08.189Z',
  generatePower: 150,
};

export const generationsWithPageMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 2,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  generations: [
    {
      id: 1,
      date: '2022-06-05T17:53:08.189Z',
      generatePower: 150,
    },
  ],
};
