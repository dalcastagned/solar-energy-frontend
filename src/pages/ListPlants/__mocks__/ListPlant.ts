export const plantsMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 1,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  plants: [
    {
      id: 1,
      nickname: 'nickname',
      place: 'place',
      brand: 'brand',
      model: 'model',
      active: true,
    },
  ],
};

export const plantMock = {
  id: 1,
  nickname: 'Apelido da Planta',
  place: 'Local da Planta',
  brand: 'Marca da Planta',
  model: 'Modelo da Planta',
  active: true,
};

export const plantsWithPageMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 2,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  plants: [
    {
      id: 1,
      nickname: 'nickname2',
      place: 'place2',
      brand: 'brand2',
      model: 'model2',
      active: false,
    },
  ],
};
