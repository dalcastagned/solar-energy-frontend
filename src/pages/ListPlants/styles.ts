import {
  Alert,
  Button,
  Card,
  TableCell,
  TableHead,
  TableRow,
  TextField,
  Box,
  CardActions,
} from '@mui/material';

import { styled } from '@mui/material/styles';

export const Wrapper = styled('div')(() => ({
  display: 'flex',
  width: '100%',
  alignItems: 'flex-end',
  flexDirection: 'column',
  gap: '3rem',
}));

export const TableHeadStyled = styled(TableHead)(() => ({
  backgroundColor: '#374557',
}));

export const TableCellTopStyled = styled(TableCell)(() => ({
  color: '#fff',
  fontSize: '1.6rem',
}));

export const TableRowStyled = styled(TableRow)(() => ({
  '&:last-child td, &:last-child th': { border: 0 },
  backgroundColor: '#fafafa',
  '&:hover': { backgroundColor: '#eaeaea' },
}));

export const TableCellBodyStyled = styled(TableCell)(() => ({
  fontSize: '1.3rem',
  cursor: 'pointer',
}));

export const ButtonEdit = styled(Button)(() => ({
  fontSize: '1.3rem',
  backgroundColor: '#8DB51B',
  color: '#fff',
  height: '3.5rem',
  width: '7.8rem',
  borderRadius: '0.4rem',
  '&:hover': {
    backgroundColor: '#7AAE1B',
  },
}));

export const ButtonRemove = styled(Button)(() => ({
  fontSize: '1.3rem',
  backgroundColor: '#D82D56',
  color: '#fff',
  height: '3.5rem',
  width: '10rem',
  borderRadius: '0.4rem',
  '&:hover': {
    backgroundColor: '#B61D4B',
  },
}));

export const ButtonNewPlant = styled(Button)(() => ({
  fontSize: '1.6rem',
  backgroundColor: '#4C5DF1',
  color: '#fff',
  height: '5.6rem',
  width: '25rem',
  borderRadius: '0.4rem',
  '&:hover': {
    backgroundColor: '#3C4DF0',
  },
}));

export const ContainerTable = styled('div')(() => ({
  display: 'flex',
  width: '100%',
  alignItems: 'flex-end',
  flexDirection: 'column',
  gap: '3rem',

  '@media screen and (max-width: 1023px)': {
    display: 'none',
  },
}));

export const AlertStyled = styled(Alert)(() => ({
  fontSize: '1.6rem',
  width: '100%',
}));

export const Loading = styled('div')(() => ({
  display: 'flex',
  width: '100%',
  height: 'calc(100vh - 10.4rem)',
  alignItems: 'center',
  justifyContent: 'center',
  flexDirection: 'column',
}));

export const TextFieldStyled = styled(TextField)(() => ({
  width: '100%',
  maxWidth: '35rem',
  label: { fontSize: '1.6rem' },
  input: { fontSize: '1.6rem' },
  '& p': { fontSize: '1.2rem' },
  '& .MuiSvgIcon-root': { fontSize: '2.25rem' },
}));

export const Form = styled('form')(() => ({
  width: '100%',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'left',
  gap: '3rem',
  alignItems: 'center',
  flexWrap: 'wrap',

  '@media screen and (max-width: 1023px)': {
    justifyContent: 'center',
  },
}));

export const ButtonSearch = styled(Button)(() => ({
  fontSize: '1.6rem',
  backgroundColor: '#4C5DF1',
  color: '#fff',
  height: '5.6rem',
  width: '15rem',
  borderRadius: '0.4rem',
  '&:hover': {
    backgroundColor: '#3C4DF0',
  },
}));

export const CardStyled = styled(Card)(() => ({
  minWidth: 260,
  backgroundColor: '#fafafa',
}));

export const ContainerCard = styled('div')(() => ({
  width: '100%',
  display: 'flex',
  flexDirection: 'row',
  flexWrap: 'wrap',
  gap: '3rem',
  alignItems: 'center',
  justifyContent: 'center',

  '@media screen and (min-width: 1023px)': {
    display: 'none',
  },
}));

export const BoxStyled = styled(Box)(() => ({
  display: 'flex',
  alignItems: 'center',
  gap: '1rem',
}));

export const ButtonEditStyled = styled(Button)(() => ({
  '& .MuiSvgIcon-root': { fontSize: '2.25rem', color: '#8DB51B' },
}));

export const ButtonRemoveStyled = styled(Button)(() => ({
  '& .MuiSvgIcon-root': { fontSize: '2.25rem', color: '#B61D4B' },
}));

export const CardActionsStyled = styled(CardActions)(() => ({
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'space-between',
  alignItems: 'center',
  gap: '1rem',
}));
