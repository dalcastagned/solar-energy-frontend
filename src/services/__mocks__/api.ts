import axios from 'axios';

export type Error = {
  status?: number;
  data?: Record<string, unknown>;
  message?: string;
};

export type HttpResponse<DataType> = {
  data?: DataType;
  error?: Error;
};

const api = jest.genMockFromModule('axios') as jest.MockedFunction<
  typeof axios
>;

api.create = jest.fn(() => api);

export default api;
