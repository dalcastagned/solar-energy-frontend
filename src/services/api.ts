import axios, { AxiosError, AxiosResponse } from 'axios';

// import { cleanStorage } from '../hooks/auth';

export type Error = {
  status?: number;
  data?: Record<string, unknown>;
  message?: string;
};

export type HttpResponse<DataType> = {
  data?: DataType;
  error?: Error;
};

export type ErrorResponse = {
  response: Error;
  code: string;
};

const responseSuccessHandler = (response: AxiosResponse): AxiosResponse =>
  response;

const forbidden = [401, 403];

const responseErrorHandler = (error: AxiosError): Promise<AxiosError> => {
  const isUnauthorized =
    error.response && forbidden.includes(error.response.status);
  const routers = ['/', '/reset-password'].includes(window.location.pathname);
  if (!routers && isUnauthorized) {
    // cleanStorage();
    window.location.href = '/';
  }

  return Promise.reject(error);
};

const api = axios.create({ baseURL: 'http://localhost:7140/api' });

api.interceptors.response.use(
  response => responseSuccessHandler(response),
  error => responseErrorHandler(error),
);

export default api;
