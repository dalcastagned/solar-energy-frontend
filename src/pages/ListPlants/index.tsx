import { useCallback, useEffect, useState } from 'react';

import DeleteIcon from '@mui/icons-material/Delete';
import EditIcon from '@mui/icons-material/Edit';
import SearchIcon from '@mui/icons-material/Search';
import {
  AlertProps,
  CardContent,
  CircularProgress,
  InputAdornment,
  Pagination,
  Typography,
} from '@mui/material';
import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableRow from '@mui/material/TableRow';
import { useFormik } from 'formik';
import { ReadPlants } from 'models/plants';
import AddPlantModal from 'pages/modals/AddPlant';
import UpdatePlantModal from 'pages/modals/UpdatePlant';
import { useHistory } from 'react-router-dom';
import { toast } from 'react-toastify';
import api from 'services/api';
import { handleErrorResponse } from 'utils/error-response';
import { notify } from 'utils/toast-notification';
import * as Yup from 'yup';

import * as S from './styles';

const validationSchema = Yup.object({
  filter: Yup.string(),
});

const Cols = [
  { id: 1, name: 'ID' },
  { id: 2, name: 'Apelido' },
  { id: 3, name: 'Local' },
  { id: 4, name: 'Marca' },
  { id: 5, name: 'Modelo' },
  { id: 6, name: 'Status' },
  { id: 7, name: '' },
  { id: 8, name: '' },
];

type ShowAlertProps = AlertProps & {
  message: string;
};

const ListPlants = (): JSX.Element => {
  const [plants, setPlants] = useState<ReadPlants>();
  const [page, setPage] = useState(1);
  const [filter, setFilter] = useState('');
  const [loading, setLoading] = useState(false);
  const [updatePlantId, setUpdatePlantId] = useState(0);
  const [openAddPlantModal, setOpenAddPlantModal] = useState(false);
  const [openUpdatePlantModal, setOpenUpdatePlantModal] = useState(false);
  const [updateTable, setUpdateTable] = useState(false);
  const [alertProps, setAlertProps] = useState<ShowAlertProps>({
    severity: 'success',
    message: '',
  });
  const history = useHistory();

  const formik = useFormik({
    initialValues: {
      filter: '',
    },
    validationSchema,
    onSubmit: values => {
      setFilter(values.filter);
    },
  });

  const getPlants = useCallback(async () => {
    setAlertProps({
      severity: 'info',
      message: '',
    });
    const response = await api.get<ReadPlants>(
      `/plant?page=${page}&limit=8&filter=${filter}`,
    );
    setLoading(false);
    !response.data.plants.length &&
      setAlertProps({
        severity: 'info',
        message: 'Nenhuma unidade cadastrada',
      });
    setPlants(response.data);
    setUpdateTable(false);
  }, [page, filter]);

  const removePlant = useCallback(
    async (id: number): Promise<void> => {
      const removeToast = toast.loading('Removendo...');
      try {
        await api.delete<ReadPlants>(`/plant/${id}`);
        notify(removeToast, {
          message: 'Planta removida com sucesso',
          type: 'success',
        });
        getPlants().catch(e =>
          setAlertProps({ severity: 'error', message: handleErrorResponse(e) }),
        );
      } catch (e) {
        notify(removeToast, {
          message: handleErrorResponse(e),
          type: 'error',
        });
      }
    },
    [getPlants],
  );

  const handleEdit = (id: number): void => {
    setUpdatePlantId(id);
    setOpenUpdatePlantModal(true);
  };

  const handleGoToGenerations = (
    id: number,
    plantName: string,
    active: boolean,
  ): void => {
    history.push('/generations', { id, plantName, active });
  };

  useEffect(() => {
    getPlants().catch(e =>
      setAlertProps({ severity: 'error', message: handleErrorResponse(e) }),
    );
  }, [getPlants, page, loading, updateTable]);

  return (
    <>
      {!alertProps.message.length && !plants?.plants.length ? (
        <S.Loading>
          <CircularProgress />
        </S.Loading>
      ) : (
        <S.Wrapper>
          <S.Form
            onSubmit={e => {
              e.preventDefault();
              formik.handleSubmit(e);
            }}
          >
            <S.TextFieldStyled
              id="filter"
              name="filter"
              InputProps={{
                placeholder: 'Apelido, Local, Marca, Modelo',
                startAdornment: (
                  <InputAdornment position="start">
                    <SearchIcon />
                  </InputAdornment>
                ),
              }}
              size="medium"
              value={formik.values.filter}
              onChange={formik.handleChange}
              error={formik.touched.filter && Boolean(formik.errors.filter)}
              helperText={formik.touched.filter && formik.errors.filter}
            />
            <S.ButtonSearch type="submit" onClick={() => setLoading(true)}>
              {loading ? <CircularProgress color="inherit" /> : 'Pesquisar'}
            </S.ButtonSearch>
          </S.Form>
          {alertProps.message.length ? (
            <S.AlertStyled severity={alertProps.severity}>
              {alertProps.message}
            </S.AlertStyled>
          ) : null}
          {plants?.plants.length ? (
            <>
              <S.ContainerTable>
                <TableContainer component={Paper}>
                  <Table size="medium" aria-label="a dense table">
                    <S.TableHeadStyled>
                      <TableRow>
                        {Cols.map(col => (
                          <S.TableCellTopStyled key={col.id}>
                            {col.name}
                          </S.TableCellTopStyled>
                        ))}
                      </TableRow>
                    </S.TableHeadStyled>
                    <TableBody>
                      {plants?.plants.map(plant => (
                        <S.TableRowStyled key={plant.id}>
                          <S.TableCellBodyStyled
                            width={50}
                            onClick={() =>
                              handleGoToGenerations(
                                plant.id,
                                plant.nickname,
                                plant.active,
                              )
                            }
                          >
                            {plant.id.toString().length > 5
                              ? `${plant.id.toString().substring(0, 5)}...`
                              : plant.id.toString()}
                          </S.TableCellBodyStyled>
                          <S.TableCellBodyStyled
                            onClick={() =>
                              handleGoToGenerations(
                                plant.id,
                                plant.nickname,
                                plant.active,
                              )
                            }
                          >
                            {plant.nickname[0].toUpperCase() +
                              plant.nickname.slice(1).toLowerCase()}
                          </S.TableCellBodyStyled>
                          <S.TableCellBodyStyled
                            onClick={() =>
                              handleGoToGenerations(
                                plant.id,
                                plant.nickname,
                                plant.active,
                              )
                            }
                          >
                            {plant.place[0].toUpperCase() +
                              plant.place.slice(1).toLowerCase()}
                          </S.TableCellBodyStyled>
                          <S.TableCellBodyStyled
                            onClick={() =>
                              handleGoToGenerations(
                                plant.id,
                                plant.nickname,
                                plant.active,
                              )
                            }
                          >
                            {plant.brand[0].toUpperCase() +
                              plant.brand.slice(1).toLowerCase()}
                          </S.TableCellBodyStyled>
                          <S.TableCellBodyStyled
                            onClick={() =>
                              handleGoToGenerations(
                                plant.id,
                                plant.nickname,
                                plant.active,
                              )
                            }
                          >
                            {plant.model[0].toUpperCase() +
                              plant.model.slice(1).toLowerCase()}
                          </S.TableCellBodyStyled>
                          <S.TableCellBodyStyled
                            onClick={() =>
                              handleGoToGenerations(
                                plant.id,
                                plant.nickname,
                                plant.active,
                              )
                            }
                          >
                            {plant.active ? 'Ativa' : 'Inativa'}
                          </S.TableCellBodyStyled>
                          <TableCell width={80}>
                            <S.ButtonEdit onClick={() => handleEdit(plant.id)}>
                              Editar
                            </S.ButtonEdit>
                          </TableCell>
                          <TableCell width={110}>
                            <S.ButtonRemove
                              onClick={() => removePlant(plant.id)}
                            >
                              Remover
                            </S.ButtonRemove>
                          </TableCell>
                        </S.TableRowStyled>
                      ))}
                    </TableBody>
                  </Table>
                </TableContainer>
              </S.ContainerTable>
              <S.ContainerCard>
                {plants?.plants.map(plant => (
                  <S.CardStyled key={plant.id}>
                    <CardContent
                      onClick={() =>
                        handleGoToGenerations(
                          plant.id,
                          plant.nickname,
                          plant.active,
                        )
                      }
                    >
                      <S.BoxStyled>
                        <Typography variant="h4">Id:</Typography>
                        <Typography variant="h4">{plant.id}</Typography>
                      </S.BoxStyled>
                      <S.BoxStyled>
                        <Typography variant="h4">Apelido:</Typography>
                        <Typography variant="h4">
                          {plant.nickname[0].toUpperCase() +
                            plant.nickname.slice(1).toLowerCase()}
                        </Typography>
                      </S.BoxStyled>
                      <S.BoxStyled>
                        <Typography variant="h4">Local:</Typography>
                        <Typography variant="h4">
                          {plant.place[0].toUpperCase() +
                            plant.place.slice(1).toLowerCase()}
                        </Typography>
                      </S.BoxStyled>
                      <S.BoxStyled>
                        <Typography variant="h4">Marca:</Typography>
                        <Typography variant="h4">
                          {plant.brand[0].toUpperCase() +
                            plant.brand.slice(1).toLowerCase()}
                        </Typography>
                      </S.BoxStyled>
                      <S.BoxStyled>
                        <Typography variant="h4">Modelo:</Typography>
                        <Typography variant="h4">
                          {plant.model[0].toUpperCase() +
                            plant.model.slice(1).toLowerCase()}
                        </Typography>
                      </S.BoxStyled>
                      <S.BoxStyled>
                        <Typography variant="h4">Status:</Typography>
                        <Typography variant="h4">
                          {plant.active ? 'Ativa' : 'Inativa'}
                        </Typography>
                      </S.BoxStyled>
                    </CardContent>
                    <S.CardActionsStyled>
                      <S.ButtonEditStyled onClick={() => handleEdit(plant.id)}>
                        <EditIcon />
                      </S.ButtonEditStyled>
                      <S.ButtonRemoveStyled
                        onClick={() => removePlant(plant.id)}
                      >
                        <DeleteIcon />
                      </S.ButtonRemoveStyled>
                    </S.CardActionsStyled>
                  </S.CardStyled>
                ))}
              </S.ContainerCard>
            </>
          ) : null}
          {plants?.plants.length ? (
            <Pagination
              count={plants?.pageCount}
              page={plants?.pageNumber}
              size="large"
              onChange={(_, value) => setPage(value)}
            />
          ) : null}
          <S.ButtonNewPlant onClick={() => setOpenAddPlantModal(true)}>
            Nova Unidade
          </S.ButtonNewPlant>
        </S.Wrapper>
      )}
      <AddPlantModal
        open={openAddPlantModal}
        setOpen={setOpenAddPlantModal}
        setUpdateTable={setUpdateTable}
      />
      <UpdatePlantModal
        id={updatePlantId}
        open={openUpdatePlantModal}
        setOpen={setOpenUpdatePlantModal}
        setUpdateTable={setUpdateTable}
      />
    </>
  );
};

export default ListPlants;
