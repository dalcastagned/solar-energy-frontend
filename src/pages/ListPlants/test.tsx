import { act, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';
import api from 'services/api';

import ListPlants from '.';
import { renderWithRoute } from '../../utils/tests/helpers';
import { plantsMock, plantsWithPageMock } from './__mocks__/ListPlant';

describe('<ListPlants />', () => {
  it('should render the main', async () => {
    const { container } = renderWithRoute(<ListPlants />);
    await waitFor(() => {
      expect(screen.getByText('Nova Unidade')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });
  it('should render alert when there are no plants', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({ data: { plants: [] } });

    renderWithRoute(<ListPlants />);
    await waitFor(() => {
      expect(
        screen.queryByText('Nenhuma unidade cadastrada'),
      ).toBeInTheDocument();
    });
  });
  it('should render alert when get plants fails', async () => {
    jest.spyOn(api, 'get').mockRejectedValue({ status: 500 });

    renderWithRoute(<ListPlants />);
    await waitFor(() => {
      expect(
        screen.queryByText(
          'O servidor parece estar indisponível. Verifique sua conexão.',
        ),
      ).toBeInTheDocument();
    });
  });
  it('should search plants', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({
      data: plantsMock,
    });

    renderWithRoute(<ListPlants />);

    await waitFor(() => {
      userEvent.type(
        screen.getByPlaceholderText('Apelido, Local, Marca, Modelo'),
        'Nome Qualquer',
      );
    });

    jest.spyOn(api, 'get').mockResolvedValue({ data: { plants: [] } });

    act(() => {
      userEvent.click(screen.getByText('Pesquisar'));
    });

    waitFor(() => {
      expect(
        screen.getByText('Nenhuma unidade cadastrada'),
      ).toBeInTheDocument();
    });
  });
  it('should go to update plant page desktop when click edit button', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({ data: plantsMock });

    const history = createMemoryHistory();
    const historyMock = jest.spyOn(history, 'push');

    renderWithRoute(
      <Router history={history}>
        <ListPlants />
      </Router>,
    );

    await waitFor(() => {
      expect(screen.getByText('ID')).toBeInTheDocument();
    });

    await act(async () => {
      userEvent.click(
        screen.getByRole('button', {
          name: 'Editar',
        }),
      );

      await waitFor(() => {
        expect(historyMock).toBeCalledWith('/plants/update', { id: 1 });
      });
    });
  });
  it('should go to update plant page mobile when click edit button', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({ data: plantsMock });

    const history = createMemoryHistory();
    const historyMock = jest.spyOn(history, 'push');

    renderWithRoute(
      <Router history={history}>
        <ListPlants />
      </Router>,
    );

    await waitFor(() => {
      expect(screen.getByText('ID')).toBeInTheDocument();
    });

    await act(async () => {
      userEvent.click(screen.getByTestId('EditIcon'));

      await waitFor(() => {
        expect(historyMock).toBeCalledWith('/plants/update', { id: 1 });
      });
    });
  });
  it('should go to add plant page when click edit button', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({ data: { plants: [] } });

    const history = createMemoryHistory();
    const historyMock = jest.spyOn(history, 'push');

    renderWithRoute(
      <Router history={history}>
        <ListPlants />
      </Router>,
    );

    await waitFor(() => {
      expect(
        screen.getByText('Nenhuma unidade cadastrada'),
      ).toBeInTheDocument();
    });

    await act(async () => {
      userEvent.click(
        screen.getByRole('button', {
          name: 'Nova Unidade',
        }),
      );

      await waitFor(() => {
        expect(historyMock).toBeCalledWith('/plants/add');
      });
    });
  });
  it('should remove plant desktop when click remove button', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({ data: plantsMock });

    jest.spyOn(api, 'delete').mockResolvedValue(jest.fn());

    renderWithRoute(<ListPlants />);

    await waitFor(() => {
      expect(screen.getByText('ID')).toBeInTheDocument();
    });

    jest.spyOn(api, 'get').mockRejectedValue({ status: 500 });

    await act(async () => {
      userEvent.click(
        screen.getByRole('button', {
          name: 'Remover',
        }),
      );

      await waitFor(() => {
        expect(screen.getByText('ID')).toBeInTheDocument();
      });
    });
  });
  it('should remove plant mobile when click remove button', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({ data: plantsMock });

    jest.spyOn(api, 'delete').mockResolvedValue(jest.fn());

    renderWithRoute(<ListPlants />);

    await waitFor(() => {
      expect(screen.getByText('ID')).toBeInTheDocument();
    });

    jest.spyOn(api, 'get').mockRejectedValue({ status: 500 });

    await act(async () => {
      userEvent.click(screen.getByTestId('DeleteIcon'));

      await waitFor(() => {
        expect(screen.getByText('ID')).toBeInTheDocument();
      });
    });
  });
  it('should go to page 2 when click page button', async () => {
    jest.spyOn(api, 'get').mockResolvedValue({ data: plantsWithPageMock });

    renderWithRoute(<ListPlants />);

    await waitFor(() => {
      expect(screen.getByText('ID')).toBeInTheDocument();
    });

    jest.spyOn(api, 'get').mockResolvedValue({ data: plantsMock });

    await act(async () => {
      userEvent.click(screen.getByText('2'));

      await waitFor(() => {
        expect(screen.getByText('ID')).toBeInTheDocument();
      });
    });
  });
  it('should render error toast on remove plant error', async () => {
    jest.useFakeTimers();

    jest.spyOn(api, 'get').mockResolvedValue({ data: plantsMock });

    jest.spyOn(api, 'delete').mockRejectedValue({ status: 500 });

    renderWithRoute(
      <div>
        <ToastContainer />
        <ListPlants />
      </div>,
    );

    await waitFor(() => {
      expect(screen.getByText('ID')).toBeInTheDocument();
    });

    userEvent.click(
      screen.getByRole('button', {
        name: 'Remover',
      }),
    );

    await waitFor(() => {
      expect(screen.queryAllByText('Removendo...')).toHaveLength(2);
    });

    await waitFor(() => {
      expect(
        screen.getByText(
          'O servidor parece estar indisponível. Verifique sua conexão.',
        ),
      ).toBeInTheDocument();
    });
  });
});
