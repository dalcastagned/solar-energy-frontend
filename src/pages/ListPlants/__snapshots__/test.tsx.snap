// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListPlants /> should render the main 1`] = `
<div>
  <div
    class="css-1q6wntd"
  >
    <form
      class="css-6sihv2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-14lzurj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
            id="filter"
            name="filter"
            placeholder="Apelido, Local, Marca, Modelo"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-hdw1oc"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-1b2e0bq-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Pesquisar
      </button>
    </form>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-1hcet0e-MuiPaper-root-MuiAlert-root"
      role="alert"
    >
      <div
        class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="ErrorOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message css-acap47-MuiAlert-message"
      >
        O servidor parece estar indisponível. Verifique sua conexão.
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-19hf11k-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Nova Unidade
    </button>
  </div>
</div>
`;
